demo-circuit/src/blaze_csw/constraints/data_structures.rs

```diff
@@ -0,0 +1,1833 @@
+            &output_hash_g,
+        )?;
+
+        // require(last_wcert.proof_data.scb_new_mst_root == mst_root)
+        mst_root_g.conditional_enforce_equal(
+            cs.ns(|| "last_wcert.proof_data.scb_new_mst_root == mst_root"),
+            scb_new_mst_root_g,
+            should_enforce,
+        )?;
+
+        // 2 Enforce nullifier
+        output_hash_g.conditional_enforce_equal(
+            cs.ns(|| "require(nullifier == outputHash)"),
+            nullifier_g,
+            should_enforce,
+        )?;
+
+        // 3. Enforce amount
+        let mut utxo_amount_big_endian_bits_g = self.input_g.output_g.amount_g.to_bits_le();
+        utxo_amount_big_endian_bits_g.reverse();
+
+        let utxo_input_amount_g = FieldElementGadget::from_bits(
+            cs.ns(|| "read utxo input amount"),
+            &utxo_amount_big_endian_bits_g,
+        )?;
+
+        utxo_input_amount_g.conditional_enforce_equal(
+            cs.ns(|| "input.amount == sys_data.amount"),
+            amount_g,
+            should_enforce,
+        )?;
+
+        Ok(())
+    }
+}
+
+impl AllocGadget<CswUtxoProverData, FieldElement> for CswUtxoProverDataGadget {
+    fn alloc<F, T, CS: ConstraintSystemAbstract<FieldElement>>(
+        mut cs: CS,
+        f: F,
+    ) -> Result<Self, SynthesisError>
+    where
+        F: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<CswUtxoProverData>,
+    {
+        let (input, mst_path_to_output) = match f() {
+            Ok(csw_utxo_prover_data) => {
+                let csw_utxo_prover_data = csw_utxo_prover_data.borrow().clone();
+                (
+                    Ok(csw_utxo_prover_data.input),
+                    Ok(csw_utxo_prover_data.mst_path_to_output),
+                )
+            }
+            _ => (
+                Err(SynthesisError::AssignmentMissing),
+                Err(SynthesisError::AssignmentMissing),
+            ),
+        };
+
+        let input_g = CswUtxoInputDataGadget::alloc(cs.ns(|| "alloc input"), || input)?;
+
+        let mst_path_to_output_g =
+            GingerMHTBinaryGadget::alloc(cs.ns(|| "alloc mst path to output"), || {
+                mst_path_to_output
+            })?;
+
+        Ok(Self {
+            input_g,
+            mst_path_to_output_g,
+        })
+    }
+
+    fn alloc_input<F, T, CS: ConstraintSystemAbstract<FieldElement>>(
+        _cs: CS,
+        _f: F,
+    ) -> Result<Self, SynthesisError>
+    where
+        F: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<CswUtxoProverData>,
+    {
+        unimplemented!()
+    }
+}
+
+#[derive(PartialEq, Eq)]
+pub struct CswFtOutputDataGadget {
+    pub amount_g: UInt64,
+    pub receiver_pub_key_g: [UInt8; SC_PUBLIC_KEY_LENGTH],
+    pub payback_addr_data_hash_g: [Boolean; MC_RETURN_ADDRESS_BYTES * 8],
+    pub tx_hash_g: [Boolean; FIELD_SIZE * 8],
+    pub out_idx_g: UInt32,
+}
+
+impl CswFtOutputDataGadget {
```

Comment:
We can apply the same optimization as discussed in the previous comment to decide whether it's phantom or not. E.g. let's take payback_addr_data_hash_g as discriminant

Reply:
Done with [commit](https://github.com/HorizenOfficial/zendoo-sc-cryptolib/commit/5bae454eef88900b13e9ec316cd567ba07d13885).